src/useIntersectionObserver.tsx

```diff
@@ -0,0 +1,34 @@
+import { RefObject, useEffect, useState } from 'react';
+
+interface Args extends IntersectionObserverInit {
+  freezeOnceVisible?: boolean;
+}
+
+function useIntersectionObserver(
+  elementRef: RefObject<Element>,
+  {
+    threshold = 0,
+    root = null,
+    rootMargin = '0%',
+    freezeOnceVisible = false,
```

Comment:
This functionality is not required.

Over-engineering is considered a bad practice. Check Google guides for details: https://google.github.io/eng-practices/review/reviewer/looking-for.html

> The developer isn’t implementing things they might need in the future but don’t know they need now.

> A particular type of complexity is over-engineering, where developers have made the code more generic than it needs to be, or added functionality that isn’t presently needed by the system. Reviewers should be especially vigilant about over-engineering. Encourage developers to solve the problem they know needs to be solved now, not the problem that the developer speculates might need to be solved in the future. The future problem should be solved once it arrives and you can see its actual shape and requirements in the physical universe.